Introduce Preliminary Guideline Document for Managing Java Key Stores and Certificate Trust Stores (#15)
wso2/container-guide: deploy/Managing_Keystores_And_Truststores.md

WSO2 products use asymmetric encryption by default for the authentication and protection of data. For advanced details,
please refer the official WSO2 [documentation](https://docs.wso2.com/display/ADMIN44x/Using+Asymmetric+Encryption#UsingAsymmetricEncryption-setting_up_keystoresUsageofkeystoresinWSO2products).

During this process, [Java keystores and truststores](https://docs.wso2.com/display/ADMIN44x/Using+Asymmetric+Encryption#UsingAsymmetricEncryption-Understandingkeystoresandtruststores)
are used.

Please refer the official WSO2 [recommendations](https://docs.wso2.com/display/ADMIN44x/Using+Asymmetric+Encryption#UsingAsymmetricEncryption-recommendationsRecommendationsforsettingupkeystoresinWSO2products)
for managing keystores prior to following this guide.

### Manage Keystores and Certificate Truststores in Container Based Deployments

As per aforementioned WSO2 [recommendations](https://docs.wso2.com/display/ADMIN44x/Using+Asymmetric+Encryption#UsingAsymmetricEncryption-recommendationsRecommendationsforsettingupkeystoresinWSO2products),
it is often required to customize the default, WSO2 product keystore and truststore in a production grade deployment.

> The most appropriate approach for introducing a custom keystore or a truststore to a container based deployment is via

Reviewer: is by using

volume mounting.

1. Prepare the custom Java keystore.

   Please follow the official guidelines for [creating a new keystore](https://docs.wso2.com/display/ADMIN44x/Creating+New+Keystores#CreatingNewKeystores-Creatinganewkeystore)
   or [adding a Certificate Authority signed certificate to an existing keystore](https://docs.wso2.com/display/ADMIN44x/Creating+New+Keystores#CreatingNewKeystores-ca_certificateAddingCA-signedcertificatestokeystores)
   as per your preference.
   

>    Also, please follow the official [documentation](https://docs.wso2.com/display/ADMIN44x/Creating+New+Keystores#CreatingNewKeystores-Step3:Importingcertificatestothetruststore)

Reviewer: To import certificates to a truststore follow this [documentation](https://docs.wso2.com/display/ADMIN44x/Creating+New+Keystores#CreatingNewKeystores-Step3:Importingcertificatestothetruststore)

   for guidelines required to import a certificate to a truststore.
   
2. Add the keystore to the product container via a volume mount depending on the container platform.

   We recommend you to use the volume mount point allocated for non-configuration resources in WSO2 product containers
   (i.e. `/home/wso2carbon/wso2-artifact-volume`) for this purpose.
   

>    In Carbon Kernel version 4 based products, the default `wso2carbon.jks` keystore is located at the

Reviewer: give a highlight to **version 4** for this point and next to help skim and scan

   `<WSO2_PRODUCT_HOME>/repository/resources/security` directory. Thus, you may mount the volume containing the
   custom keystore to `/home/wso2carbon/wso2-artifact-volume/repository/resources/security` directory.
   

>    In Carbon Kernel version 5 based products, the default `wso2carbon.jks` keystore is located at the

Reviewer: **version 5**

   `<WSO2_PRODUCT_HOME>/resources/security` directory. Thus, you may mount the volume containing the custom keystore
   to `/home/wso2carbon/wso2-artifact-volume/resources/security` directory.
   
3. Add the truststore to the product container via a volume mount depending on the container platform.

   We recommend you to use the volume mount point allocated for non-configuration resources in WSO2 product containers
   (i.e. `/home/wso2carbon/wso2-artifact-volume`) for this purpose.
   

>    In Carbon Kernel version 4 based products, the default `client-truststore.jks` truststore is located at the

Reviewer: Same as before for this and the other point

   `<WSO2_PRODUCT_HOME>/repository/resources/security` directory. Thus, you may mount the volume containing the
   custom keystore to `/home/wso2carbon/wso2-artifact-volume/repository/resources/security` directory.
   
   In Carbon Kernel version 5 based products, the default `client-truststore.jks` truststore is located at the
   `<WSO2_PRODUCT_HOME>/resources/security` directory. Thus, you may mount the volume containing the custom keystore
   to `/home/wso2carbon/wso2-artifact-volume/resources/security` directory.
    

>    > In Kubernetes based WSO2 product deployments, we recommend you to use a ConfigMap resource to add the custom keystore

Reviewer: product deployment, it is advisable that you used
```suggestion
   > In Kubernetes based WSO2 product deployment, it is advisable that you use a ConfigMap resource to add the custom keystore
```

   > and truststore.
   > 

>    > * [Create a ConfigMap resource from the keystore file](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-files)

Reviewer: ```suggestion
   > * For further understanding peruse:
   

- [Create a ConfigMap resource from the keystore file](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-files)

```

   `<WSO2_PRODUCT_HOME>/resources/security` directory. Thus, you may mount the volume containing the custom keystore
   to `/home/wso2carbon/wso2-artifact-volume/resources/security` directory.
   
3. Add the truststore to the product container via a volume mount depending on the container platform.

   We recommend you to use the volume mount point allocated for non-configuration resources in WSO2 product containers
   (i.e. `/home/wso2carbon/wso2-artifact-volume`) for this purpose.
   
   In Carbon Kernel **version 4** based products, the default `client-truststore.jks` truststore is located at the
   `<WSO2_PRODUCT_HOME>/repository/resources/security` directory. Thus, you may mount the volume containing the
   custom keystore to `/home/wso2carbon/wso2-artifact-volume/repository/resources/security` directory.
   
   In Carbon Kernel **version 5** based products, the default `client-truststore.jks` truststore is located at the
   `<WSO2_PRODUCT_HOME>/resources/security` directory. Thus, you may mount the volume containing the custom keystore
   to `/home/wso2carbon/wso2-artifact-volume/resources/security` directory.
    

>    > In Kubernetes based WSO2 product deployment, it is advisable that you use a ConfigMap resource to add the custom keystore

Reviewer 2: Would it be better if we mention the Kubernetes versions that support this approach? AFAIR it is from `1.10.0` right?.

Author: @ThilinaManamgoda +1. Will add.

Author: Added in https://github.com/wso2/container-guide/pull/15/files#diff-9bf3730dcb649a313efa880c13afc382R69.
